loc-analysis.exs

```diff
@@ -0,0 +1,202 @@
+#! /usr/bin/env elixir
+
+Mix.install([
+  {:csv, "~> 3.0"}
+])
+
+defmodule GitCommandError do
+  defexception message: "Git command failed."
+
+  @impl true
+  def exception(value) do
+    case value do
+      [] -> %GitCommandError{}
+      _ -> %GitCommandError{message: "Git command failed: " <> value}
+    end
+  end
+end
+
+defmodule Git do
+  def get_snapshot!(commit_hash) do
+    with destination_path <- System.tmp_dir!(),
+         {:ok, tar_file} <- call_git_archive!(commit_hash, destination_path),
+         :ok <- call_tar_extract!(tar_file, destination_path),
+         :ok <- Path.join(destination_path, tar_file) |> File.rm!() do
+      {:ok, Path.join(destination_path, commit_hash)}
+    end
+  end
+
+  def monthly_commits do
+    call_git_log!()
+    |> String.split("\n")
+    |> Stream.uniq_by(&year_month_of/1)
+  end
+
+  def weekly_commits do
+    call_git_log!()
+    |> String.split("\n")
+    |> Stream.uniq_by(&beginning_of_week/1)
+  end
+
+  def call_git_log! do
+    case System.cmd("git", [
+           "log",
+           "--pretty=format:%ad,%H",
+           "--date=iso",
+           "--date=short",
+           "--reverse"
+         ]) do
+      {log, 0} -> log
+      {error_msg, _} -> raise GitCommandError, error_msg
+    end
+  end
+
+  def parse_git_log_line(log_line), do: log_line |> String.split(",")
+
+  defp year_month_of(commit), do: commit |> String.slice(0..6)
+
+  defp beginning_of_week(commit),
+    do:
+      commit
+      |> String.slice(0..9)
+      |> Date.from_iso8601()
+      |> elem(1)
+      |> Date.beginning_of_week()
+      |> Date.to_string()
+
+  defp call_git_archive!(commit_hash, temp_dir_path_str) do
+    System.cmd("git", [
+      "archive",
+      commit_hash,
+      "--prefix=#{commit_hash}/",
+      "--format=tar",
+      "--output=#{temp_dir_path_str}/#{commit_hash}.tar"
+    ])
+    |> case do
+      {_, 0} -> {:ok, "#{commit_hash}.tar"}
+    end
+  end
+
+  defp call_tar_extract!(archive_file, archive_location) do
+    System.cmd(
+      "tar",
+      [
+        "xf",
+        archive_file
+      ],
+      cd: archive_location
+    )
+    |> case do
+      {_, 0} -> :ok
+    end
+  end
+end
+
+defmodule Analysis do
+  def monthly(file) do
+    Git.monthly_commits()
```

Comment:
Now _this_ function packs a _lot_ of work in a few lines. Nothing bad with that, I'd just consider splitting the work a bit. One concrete idea: the function you pass to `map_reduce` can be implemented elsewhere.